### DISCUSSION TITLE
vearn.finance

### DISCUSSION THREAD

**Discussion by fede-rodes** [Cyfrin/security-and-auditing-full-course-s23#136] [Show and tell]
Hi everyone!

I'm building a dApp on VeChain and would love your feedback! Both from the UX perspective and smart contract implementation. The app is called Vearn and it's deployed on testnet at https://vearn.finance/.

**Context**
VeChain operates as an EVM-compatible network with a unique two-token model:

- **VTHO**: An ERC20 token primarily used as gas.
- **VET**: The native token, which generates VTHO at a consistent rate of 5*10^-8 VTHO per VET per block when held in an account or contract.

**Idea:**
The protocol aims to automate the exchange of generated VTHO tokens for additional VET tokens by employing a mathematical approach to determine the "optimal" timing for the swap. This optimized timing is calculated off-chain. When determined, a call to the Trader smart contract (outlined below) is made to initiate the token swap via a decentralized exchange (DEX).

You can find the full source code of the contract here: https://github.com/vearnfi/contracts/blob/main/contracts/Trader.sol

Or view it displayed in a nice format here: https://carbon.now.sh/Ua4EVQLrmTjd46tqD58q

Thanks so much for your time and also to all the Cyfrin team for providing all the amazing content! 🙌

**Comment by EngrPips** [comment]
Hey @fede-rodes  thats a great courage and motivation to actually build something out yourself. Trust anyone that is chanced would look into your protocol and give feedback as needed.

**Reply by fede-rodes** [reply]
Thanks a lot!
